src/daft-minhash/src/windowed.rs

```diff
@@ -0,0 +1,319 @@
+struct WindowedWords<'a> {
+    s: &'a str,
+    word_starts: Vec<usize>, // Vec of start indices for each word
+    window_size: usize,
+    current: usize, // Current starting word index for the window
+}
+
+pub trait WindowedWordsExt<'a> {
+    fn windowed_words(&'a self, window_size: usize) -> impl Iterator<Item = &'a str>;
+}
+
+impl<'a> WindowedWordsExt<'a> for str {
+    fn windowed_words(&'a self, window_size: usize) -> impl Iterator<Item = &'a Self> {
+        WindowedWords::new(self, window_size)
+    }
+}
+
+impl<'a> WindowedWords<'a> {
+    /// Creates a new `WindowedWords` iterator.
+    ///
+    /// # Arguments
+    ///
+    /// * `s` - The input string slice.
+    /// * `window_size` - The number of words in each window.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// let s = "The quick brown fox";
+    /// let iter = WindowedWords::new(s, 2);
+    /// ```
+    pub fn new(s: &'a str, window_size: usize) -> Self {
+        assert!(window_size > 0, "Window size must be greater than 0");
+
+        if s.is_empty() {
+            return WindowedWords {
+                s,
+                word_starts: vec![],
+                window_size,
+                current: 0,
+            };
+        }
+
+        // assume first character is not whitespace
+        let mut word_starts = vec![0];
+
+        for (i, _) in s.match_indices(' ') {
+            // assume character after whitespace is not whitespace
+            word_starts.push(i + 1);
+        }
+
+        WindowedWords {
+            s,
+            word_starts,
+            window_size,
+            current: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for WindowedWords<'a> {
+    type Item = &'a str;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.window_size == 0 {
+            return None;
+        }
+
+        if self.current + self.window_size <= self.word_starts.len() {
+            // Get the start of the current window
+            let start = self.word_starts[self.current];
```

Comment:
can just pluck from `match_indices` iterator directly

Reply:
We would need two `match_indices` iterators to do this. Is this something that we want to have?

Reply:
![image](https://github.com/user-attachments/assets/1a6d050d-e816-4d70-993b-520ff919c3f5)

my microbenchmarks when using two match indices (compared to previous method)

Reply:
Which one is which?

Also why do we need two iterators, can we have something like a fixed size VecDeque stores the indices between the start of the current window and the end? We can likely reuse the same VecDeque across invokations of the the minhash function.

Reply:
![image](https://github.com/user-attachments/assets/78b4f6c0-1ef3-43b9-9f75-02aa9d720d86)

ok got this (still with allocating) though. are we happy with this?

Reply:
![image](https://github.com/user-attachments/assets/658d41b4-a87a-4a62-8dc6-bd0b33a53813)
improved the empty_string case

Reply:
Before we allocating once per "row" of the series. If we can pass in a `&mut VecDeque` that is size of `window_length` we should only have 1 allocation for the whole function! That should be acceptable